0003-erl_child_setup_thread.patch

```diff
@@ -0,0 +1,1219 @@
++        wd = NULL;
++    } else {
++        wd = buff;
++        buff += strlen(buff) + 1;
++    }
++    buff++;
++
++    DEBUG_PRINT("wd = %s", wd);
++
++    cnt = get_int32(buff);
++    buff += sizeof(cnt);
++    new_environ = malloc(sizeof(char*)*(cnt + 1));
++
++    DEBUG_PRINT("env_len = %d", cnt);
++    for (i = 0; i < cnt; i++, buff++) {
++        new_environ[i] = buff;
++        while(*buff != '\0') buff++;
++    }
++    new_environ[cnt] = NULL;
++
++    if (o_buff + size != buff) {
++        /* This is a spawn executable call */
++        cnt = get_int32(buff);
++        buff += sizeof(cnt);
++        args = malloc(sizeof(char*)*(cnt + 1));
++        for (i = 0; i < cnt; i++, buff++) {
++            args[i] = buff;
++            while(*buff != '\0') buff++;
++        }
++        args[cnt] = NULL;
++    }
++
++    if (o_buff + size != buff) {
++        errno = EINVAL;
++        errln = __LINE__;
++        warning("erl_child_setup: failed with protocol "
++                "error %d on line %d", errno, errln);
++        /* we abort here as it is most likely a symptom of an
++           emulator/erl_child_setup bug */
++        abort();
++    }
++
++    DEBUG_PRINT("read ack");
++    do {
++        ErtsSysForkerProto proto;
++        res = read(pipes[0], &proto, sizeof(proto));
++        if (res > 0) {
++            ASSERT(proto.action == ErtsSysForkerProtoAction_Ack);
++            ASSERT(res == sizeof(proto));
++        }
++    } while(res < 0 && (errno == EINTR || errno == ERRNO_BLOCK));
++
++    DEBUG_PRINT("... read res %d\n", res);
++    if (res < 1) {
++        errno = EPIPE;
++        errln = __LINE__;
++        goto child_error;
++    }
++
++    DEBUG_PRINT("Set cwd to: '%s'",cwd);
++
++    if (chdir(cwd) < 0) {
++        /* This is not good, it probably means that the cwd of
++           beam is invalid. We ignore it and try anyways as
++           the child might now need a cwd or the chdir below
++           could take us to a valid directory.
++        */
++    }
++
++    DEBUG_PRINT("Set wd to: '%s'",wd);
++
++    if (wd && chdir(wd) < 0) {
++        int err = errno;
++        fprintf(stderr,"spawn: Could not cd to %s\r\n", wd);
++        _exit(err);
++    }
++
++    DEBUG_PRINT("Do that forking business: '%s'",cmd);
++
++    /* simple dispatch */
++    p = strchr(cmd, ' ');
++    if (p) {
++        *p = '\0';
++        if (strcmp(cmd, "exec")) {
++            warning("not exec command; unhandled: \"%s %s\"\n", cmd, p + 1);
++        } else {
++            cmd = p + 1;
++            p = strchr(cmd, ' ');
++            if (p)
++                *p = '\0';
++            /* ignore args for now... */
++            DEBUG_PRINT("program name is \"%s\"", cmd);
++            if (!strcmp(cmd, "inet_gethost") && !inet_gethost_started) {
++                inet_gethost_started = 1;
++                launched = 1;
++                if (pthread_create(&inet_gethost_thread, 0, simple_inet_gethost_thread, pipes) < 0) {
```

Comment:
If the runtime tries to exec inet_gethost more than once, using different pipe file descriptors each time, then exec requests after the first one won't be honored (the pipes will be closed right away). It seems simpler to just invoke simple_inet_gethost with a direct function call instead of starting another thread.

Reply:
From my understanding, inet_gethost is intended to be run as a background service to continuously receive requests over the established pipes. I have not observed the "exec inet_gethost" occurring more than once, and I do see the single thread fielding multiple requests during runtime. That said, I don't see the harm in allowing another inet_gethost to be started (which looks to be possible if lib/kernel/src/inet_gethost_native.erl:restart_port() is called, for example if there are errors on parsing a response from inet_gethost, and also possibly as a result of a "soft_restart" issued via the control API). I removed the inet_gethost_started flag and now just allow multiple instances to be run.

I agree the added pthread is not necessary, as the pthread created for start_new_child() can become the background inet_gethost thread. I changed this to call simple_inet_gethost() as you suggested. Additionally, the pthread_t for each child thread created in erl_child_setup_thread() is stored in the corresponding ErtsSysForkerProto (as os_pthread, which is now in a union with os_pid) for later use. I also added prctl() calls to set the name of all created threads to give more context when taking traces.